### PR TITLE
Chat ACS Adapter as NPM dependency <<React Native tested>>

### PR DESCRIPTION
**Pull request by elopezanaya**
# PR Details

### Description
_Include a description of the problem to be solved_

ACS Chat Adapter js is being downloaded from a host site which is not Microsoft official (unpkg).

## Solution Proposed

The switch of strategy will now use the npm dependency and load the adapter from the module, to achieve this it was required:

- add special configuration for unit tests  with separation for node and web tests, this is needed due to specific setup required for each environment for unit testing, without get in conflict durig compile time.
   - setup for jsdom (web)
   - setup for js.node (node and reactive unit tests)
   - new script commands for running test independently of all at once
   - new script command to add prebuild format validation
- Add complementary rules is eslint to solve multiple issue in format and prevent nit comments during code reviews
- Add node module for acs chat adapter
  -  since the component doesnt export its own types, it was required to use declare module to define the type, 
  -  extension to use our own type definition is not allowed since this is a blocked scope function
- no changes needed for Integration tests.

_Detail what is the solution proposed, include links to design document if required or any other document required to support the solution_

### Acceptance criteria

_Define what are the conditions to consider the PR has achieved the intended goal_

## Test cases and evidence

_Include what tests cases were considered, any evidence of testing for future references, to identify any corner cases, etc_

### OCW

image

### Unit tests

image

### Integration Authenticated 
image

### Integration non-authenticated
image

### React Native Mobile
image
image
image

### Sanity Tests

- [ ] You have tested all changes in Popout mode
- [ ] You have tested all changes in cross browsers i.e Edge, Chrome, Firefox, Safari and mobile devices(iOS and Android)
- [ ] Your changes are included in the [CHANGELOG](../CHANGE_LOG.md)

## A11y

- [ ] You have run the [Automated Accessibility Check](https://accessibilityinsights.io/docs/en/windows/getstarted/automatedchecks) and have no reported issues

***Please provide justification if any of the validations has been skipped.***

### CODE REVIEW

**Comment by dleblondMSFT** [comment]
Can you provide more context in the PR description -  I see new files for jest.setup.jsdom.js - does this mean we get UT x2 in both flavors (Node and Web)?

File: package.json [hunk 1]
```diff
@@ -36,15 +38,17 @@
...
+    "ts-jest": "^29.2.5",
+    "typescript": "^5.4.2"
   },
   "dependencies": {
     "@azure/communication-chat": "1.1.1",
     "@azure/communication-common": "1.1.0",
     "@microsoft/ocsdk": "^0.5.6",
     "@microsoft/omnichannel-amsclient": "^0.1.6",
-    "@microsoft/omnichannel-ic3core": "^0.1.4"
+    "@microsoft/omnichannel-ic3core": "^0.1.4",
```
Comment (dleblondMSFT) [inline package.json:51]: Do we really still need ic3core?

Comment (dleblondMSFT) [reply package.json:51]: I see references. We should track a separate work item to remove this ASAP.

File: src/override_type/acs_webchat-chat-adapter.d.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+// typings/acs_webchat-chat-adapter.d.ts
+declare module 'acs_webchat-chat-adapter' {
+    export function createACSAdapter(...args: any[]): any;
```
Comment (dleblondMSFT) [inline src/override_type/acs_webchat-chat-adapter.d.ts:4]: Let's fix this, shared you a typing offline.

File: __tests__/OmnichannelChatSDK.spec.ts [hunk 3]
```diff
@@ -31,12 +31,22 @@ describe('Omnichannel Chat SDK, Sequential', () => {
...
+        }
+    });
+
     describe('Configurations', () => {
         it('ChatSDK should require omnichannelConfig as parameter', () => {
             try {
                 new OmnichannelChatSDK();
                 fail();
-            } catch (error : any ) {
+            } catch (error:any ) {
```
Comment (xTEddie) [inline __tests__/OmnichannelChatSDK.spec.ts:49]: nit: missing space after `error:`

File: __tests__/OmnichannelChatSDKParallel.spec.ts [hunk 4]
```diff
@@ -27,6 +27,16 @@ describe('Omnichannel Chat SDK, Parallel initialization', () => {
         widgetId: '[data-app-id]'
     };
 
+    beforeAll(()=>{
```
Comment (xTEddie) [inline __tests__/OmnichannelChatSDKParallel.spec.ts:30]: nit: missing space after `=>`

File: __tests__/utils/WebUtils.web.spec.ts [hunk 5]
```diff
@@ -6,6 +6,16 @@
 const WebUtils = require('../../src/utils/WebUtils');
 
 describe('WebUtils', () => {
+
+    beforeEach(() => {
+        jest.resetModules();
+    }
```
Comment (xTEddie) [inline __tests__/utils/WebUtils.web.spec.ts:12]: nit: should `}` and `);` be in the same line?

File: __tests__/utils/chatAdapterCreators.web.spec.ts [hunk 6]
```diff
@@ -1,15 +1,27 @@
...
+import { ACSAdapterLogger } from "../../src/utils/loggers";
 import LiveChatVersion from "../../src/core/LiveChatVersion";
 import WebUtils from "../../src/utils/WebUtils";
 
 describe("Chat Adapter Creators", () => {
-    it ("createDirectLine() should throw an exception if script load fails", async () => {
+
+    beforeEach(() => {
+        jest.resetModules();
+    }
```
Comment (xTEddie) [inline __tests__/utils/chatAdapterCreators.web.spec.ts:17]: nit: `}` & `);` formatting

File: __tests__/utils/platform.web.spec.ts [hunk 7]
```diff
@@ -5,6 +5,16 @@
 import platform from "../../src/utils/platform";
 
 describe('Platform', () => {
+
+    beforeEach(() => {
+        jest.resetModules();
+    }
```
Comment (xTEddie) [inline __tests__/utils/platform.web.spec.ts:11]: nit: `}` & `);` formatting

File: __tests__/utils/utilities.spec.ts [hunk 8]
```diff
@@ -1,10 +1,25 @@
...
 
 /* eslint-disable @typescript-eslint/no-var-requires */
 const { MessageType } = require("../../src");
-const utilities = require('../../src/utils/utilities');
 
 describe('Utilities', () => {
+
+    beforeEach(() => {
+        jest.resetModules();
+    }
```
Comment (xTEddie) [inline __tests__/utils/utilities.spec.ts:15]: nit: `}` & `);` formatting

File: __tests__/utils/utilities.spec.ts [hunk 9]
```diff
@@ -1,10 +1,25 @@
 import { getRuntimeId, isNotEmpty } from "../../src/utils/utilities";
+import { isClientIdNotFoundErrorMessage, isCustomerMessage, isSystemMessage } from "../../src/utils/utilities";
+
+/**
+ * @jest-environment node
```
Comment (xTEddie) [inline __tests__/utils/utilities.spec.ts:5]: `@jest-environment` comment should be all the way at the top of the file

File: jest.config.jsdom.js [hunk 10]
```diff
@@ -0,0 +1,26 @@
+module.exports = {
+  preset: 'ts-jest',
+  testEnvironment: 'node',
```
Comment (xTEddie) [inline jest.config.jsdom.js:3]: Should `testEnvironment` still be `node`?

File: __tests__/utils/utilities.spec.ts (hunk 8 above)
Comment (dleblondMSFT) [reply __tests__/utils/utilities.spec.ts:15]: Can we get a task to enable VSCode autofix lint rules so we don't need manual review for this? I can share examples of repos with this setup elsewhere if needed as well.

File: jest.setup.jsdom.js [hunk 11]
```diff
@@ -0,0 +1,19 @@
+const { TextEncoder: PolyfillTextEncoder, TextDecoder: PolyfillTextDecoder } = require("text-encoding");
```
Comment (xTEddie) [inline jest.setup.jsdom.js:1]: `text-encoding` package is deprecated

File: src/utils/__mocks__/acs_webchat-chat-adapter.js [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+        console.log("end");
+    },
+    fileManager : {
+        fileScanner : {
+            end : () => {
+                console.log("file scanner end");
+            },
+        },
+    }
+};
```
Comment (xTEddie) [inline src/utils/__mocks__/acs_webchat-chat-adapter.js:16]: nit: `};` formatting issue

File: tsconfig.json [hunk 13]
```diff
@@ -7,6 +7,7 @@
         "declaration": true,
         "outDir": "./lib",
         "strict": true,
+        "typeRoots": ["./node_modules/@types", "./override_type"]
```
Comment (xTEddie) [inline tsconfig.json:10]: Would it make sense to use `./types` instead of `override_type`?

File: CHANGELOG.md [hunk 14]
```diff
@@ -4,6 +4,11 @@ All notable changes to this project will be documented in this file.
 
 ## [Unreleased]
 
+### Changed
+
+- Load ACS Chat adapter as npm dependency, instead of download it from unpk host.
```
Comment (xTEddie) [inline CHANGELOG.md:9]: nit: `unpkg`

File: __tests__/utils/utilities.spec.ts (hunk 8 above)
Comment (elopezanaya) [reply __tests__/utils/utilities.spec.ts:15]: I have added extra rules and ensure those rules are enforced when building

File: jest.config.jsdom.js (hunk 10 above)
Comment (elopezanaya) [reply jest.config.jsdom.js:3]: yeah... its default, setup function will take care of the rest

File: src/override_type/acs_webchat-chat-adapter.d.ts (hunk 2 above)
Comment (elopezanaya) [reply src/override_type/acs_webchat-chat-adapter.d.ts:4]: since we are loading now from node module, which doesnt contains its own type, it seems like the only way to avoid the error is by declaring the module.

This technique doesn allow to use our own type definition, since this is a blocked scope function, so it cant be extended. 

We can explore more in future tasks.

**Review by mithunjohny — COMMENTED** [review]
We have to wait pushing this change to prod, because acs adapter stops scrubbing pII data in telemetry in latest acs adapter version

**Review by matan-hershberg-nuance — APPROVED** [review]

**Review by matan-hershberg-nuance — APPROVED** [review]

**Review by mithunjohny — APPROVED** [review]
